Add cp and mv commands

isaacgravenortechnologies opened the pull request:
This commit/PR adds 'cp' and 'mv', where 'mv' uses the same code as 'cp' but attempts to delete the file afterwards. Command-line fkags (such as -r ), or moving/copying folders isn't supported by this code (yet!).

isaacgravenortechnologies committed:
Add cp and mv commands

This commit adds 'cp' and 'mv', where 'mv' uses the same code as 'cp' but attempts to delete the file afterwards.
Command-line fkags (such as -r ), or moving/copying folders isn't supported by this code (yet!).

iProgramMC commented on src/shell.c:
```diff
@@ -903,6 +903,71 @@ void ShellExecuteCommand(char* p, bool* pbExit)
+			FiClose(fd);
+			return;
+		}
+		else {
```
Code style.

isaacgravenortechnologies replied:
Fixed in latest commit

iProgramMC commented on src/shell.c:
```diff
@@ -903,6 +903,71 @@ void ShellExecuteCommand(char* p, bool* pbExit)
+		unsigned int sz;
+		sz = FiTell(fd);
+		FiSeek(fd, 0, SEEK_SET);
+		char* data = (char*)MmAllocate(sz+1);
```
Don't allocate the entire size of the file, instead, read it in blocks.

The `movedata` command actually does pretty much everything you need `cp` to do. You could have just renamed the `movedata` command to `cp`. Anyways this shell sucks and I need to redo it sometime. Though I probably won't because I'm working on [Boron](https://github.com/iProgramMC/Boron) right now.

isaacgravenortechnologies replied:
Modified in last commit to reuse movedata code.

iProgramMC commented on src/shell.c:
```diff
@@ -903,6 +903,71 @@ void ShellExecuteCommand(char* p, bool* pbExit)
+		char* fileName = Tokenize (&state, NULL, " ");
+
+		char* fileName2 = Tokenize (&state, NULL, " ");
+		int fd = FiOpen(fileName, O_RDONLY);
```
Check if the file names exist in the first place and aren't null or empty.

isaacgravenortechnologies replied:
Fixed in latest commit

iProgramMC commented on src/shell.c:
```diff
@@ -903,6 +903,71 @@ void ShellExecuteCommand(char* p, bool* pbExit)
+	{
+		char* fileName = Tokenize (&state, NULL, " ");
+		char* fileName2 = Tokenize (&state, NULL, " ");
+		int fd = FiOpen(fileName, O_RDONLY);
```
Again, check file names.

isaacgravenortechnologies replied:
Fixed in latest commit

iProgramMC commented on src/shell.c:
```diff
@@ -903,6 +903,71 @@ void ShellExecuteCommand(char* p, bool* pbExit)
+		unsigned int sz;
+		sz = FiTell(fd);
+		FiSeek(fd, 0, SEEK_SET);
+		char* data = (char*)MmAllocate(sz+1);
```
Same issue as `cp`.

isaacgravenortechnologies replied:
Same solution as cp.

iProgramMC commented on src/shell.c:
```diff
@@ -903,6 +903,71 @@ void ShellExecuteCommand(char* p, bool* pbExit)
+		FiWrite(fd, data, sz);
+		FiClose (fd);
+		// Get rid of the file.
+		int io = FiUnlinkFile (fileName);
```
BTW, you can use `FiRenameFile` to move a file across a file system mount point (for example, you can move `/Ext0/stuff/file.bin` to `/Ext0/otherstuff/file.bin` by just calling `FiRenameFile` on them).  First, try that call. Then, if `-ENXIO` is returned, do the copy/delete combination.  This is way faster and doesn't copy and delete a file for no reason.

isaacgravenortechnologies replied:
Done

isaacgravenortechnologies committed:
Added filename check + fixed code style

Yet to do:
block reading,
FiRenameFile

isaacgravenortechnologies committed:
Various changes

Used movedata code for cp (and for the second attempt of mv, where it copies/deletes), reusing movedata_precision for cp and mv by defining it up top.

iProgramMC commented on src/shell.c:
```diff
@@ -903,6 +902,195 @@ void ShellExecuteCommand(char* p, bool* pbExit)
+	{
+		LogMsg("Starting to copy...");
+		char* fileNameIn = Tokenize (&state, NULL, " ");
+		if (!fileNameIn)
```
One more thing. Consider doing an OR here, like, `!fileName || *fileName == 0`. I should have done that myself.

isaacgravenortechnologies replied:
Added.

iProgramMC commented on src/shell.c:
```diff
@@ -903,6 +902,195 @@ void ShellExecuteCommand(char* p, bool* pbExit)
+		size_t sz = FiTellSize(fd_in);
+		
+		//write 512 byte blocks
+		for (size_t i = 0; i < sz; i += MOVEDATA_PRECISION)
```
This seems fine. However at this point I would create a new function entirely, something like `ShellCopy(pFileNameSrc, pFileNameDst)`.

isaacgravenortechnologies replied:
I looked into doing this, but there would then need to be a consideration for which command was being run regardless, for the sake of the error messages that appear during copying if anything happens, so I held off on this idea for the moment.

iProgramMC commented on src/shell.c:
```diff
@@ -903,6 +902,195 @@ void ShellExecuteCommand(char* p, bool* pbExit)
+			return;
+		}
+		int result = FiRename(fileNameIn, fileNameOut);
+		if(result == -ENXIO) {
```
Code style. Place the `{` on a new line and place a space after the `if`.

isaacgravenortechnologies replied:
Done.

iProgramMC commented on src/shell.c:
```diff
@@ -903,6 +902,195 @@ void ShellExecuteCommand(char* p, bool* pbExit)
+		if(result == -ENXIO) {
+			LogMsg("FiRename failed, trying move+delete");
+			// Get rid of the file.
+					int fd_in = FiOpen (fileNameIn, O_RDONLY);
```
Fix whitespace.

isaacgravenortechnologies replied:
Done.

iProgramMC commented on src/shell.c:
```diff
@@ -903,6 +902,195 @@ void ShellExecuteCommand(char* p, bool* pbExit)
+			//write 512 byte blocks
+			for (size_t i = 0; i < sz; i += MOVEDATA_PRECISION)
+			{
+				size_t read_in = FiRead(fd_in, chunk_of_data, MOVEDATA_PRECISION);
```
`FiRead` returns an int.  But maybe this is my fault?

isaacgravenortechnologies replied:
`size_t` ended up being useless, since FiRead returned an int as mentioned, and everywhere it was being used, it was being casted to an `int` anyway, so I simply changed the type of `read_in` to an `int` completely (at least, for `cp` and `mv`).

isaacgravenortechnologies committed:
Merge branch 'master' of https://github.com/iProgramMC/NanoShellOS
Merge fixes from upstream

isaacgravenortechnologies committed:
Various requested changes

The code styles were all sorted out, and after having looked into other source files, size_t seems to be pointless here so int has been used instead, with casts removed where appropriate. Also, '||' has been added when checking filenames, as suggested.

isaacgravenortechnologies commented:
I believe all of the reviewed points have been addressed, save for the `ShellCopy` idea, as I commented on. Let me know what you think. 😄
